### PR TITLE
8283187: C2: loop candidate for superword not always unrolled fully if superword fails

### PR DESCRIPTION
**Pull request by rwestrel**
IdealLoopTree::policy_unroll_slp_analysis() restricts unrolling of
loops that are found to be candidate for superword but if superword
then fails, unrolling doesn't always resume. SuperWord::output() has
logic for that in one of the failure paths but some other paths are
not covered (for instance in the case of the test case). I propose
that superword explictly returns whether it succeeded or not and that
in case of failure, for a main loop, unrolling be attempted again.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283187](https://bugs.openjdk.java.net/browse/JDK-8283187): C2: loop candidate for superword not always unrolled fully if superword fails

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7822/head:pull/7822` \
`$ git checkout pull/7822`

Update a local copy of the PR: \
`$ git checkout pull/7822` \
`$ git pull https://git.openjdk.java.net/jdk pull/7822/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7822`

View PR using the GUI difftool: \
`$ git pr show -t 7822`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7822.diff">https://git.openjdk.java.net/jdk/pull/7822.diff</a>

</details>

### CODE REVIEW

File: test/hotspot/jtreg/compiler/c2/irTests/TestSuperwordFailsUnrolling.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+    private static int v = 0;
+    private final static WhiteBox wb = WhiteBox.getWhiteBox();
+
+    public static void main(String[] args) {
+        Object avx = wb.getVMFlag("UseAVX");
+        if (avx != null && ((Long)avx) > 2) {
+            TestFramework.runWithFlags("-XX:UseAVX=2");
+        }
+        TestFramework.run();
+
```
Comment (TobiHartmann) [inline test/hotspot/jtreg/compiler/c2/irTests/TestSuperwordFailsUnrolling.java:49]:
```suggestion
```

**Review by TobiHartmann — APPROVED** [review]
Good catch! The fix looks good to me.

**Comment by rwestrel** [comment]
> Good catch! The fix looks good to me.

thanks for the review! I applied your suggestions.

File: src/hotspot/share/opto/superword.cpp [hunk 2]
```diff
@@ -95,38 +95,38 @@ SuperWord::SuperWord(PhaseIdealLoop* phase) :
...
-        if (val < 0) return;
+        if (val < 0) return false;
       }
     }
   }
 
   // skip any loop that has not been assigned max unroll by analysis
   if (do_optimization) {
-    if (SuperWordLoopUnrollAnalysis && cl->slp_max_unroll() == 0) return;
+    if (SuperWordLoopUnrollAnalysis && cl->slp_max_unroll() == 0) return false;
```
Comment (chhagedorn) [inline src/hotspot/share/opto/superword.cpp:129]: While at it, you could also add braces for the one liner ifs.

**Review by chhagedorn — APPROVED** [review]
That looks good!

**Review by TobiHartmann — CHANGES_REQUESTED** [review]
I executed some testing and `compiler/c2/irTests/TestSuperwordFailsUnrolling.java` fails on Windows only with:
```
java.lang.Error: failed to install whitebox classes
	at compiler.lib.ir_framework.TestFramework.installWhiteBox(TestFramework.java:529)
	at compiler.lib.ir_framework.TestFramework.start(TestFramework.java:318)
	at compiler.lib.ir_framework.TestFramework.run(TestFramework.java:210)
	at compiler.lib.ir_framework.TestFramework.run(TestFramework.java:199)
	at compiler.c2.irTests.TestSuperwordFailsUnrolling.main(TestSuperwordFailsUnrolling.java:48)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
	at java.base/java.lang.reflect.Method.invoke(Method.java:577)
	at com.sun.javatest.regtest.agent.MainWrapper$MainThread.run(MainWrapper.java:127)
	at java.base/java.lang.Thread.run(Thread.java:828)
Caused by: java.nio.file.FileSystemException: sun\\hotspot\\WhiteBox.class: The process cannot access the file because it is being used by another process
	at java.base/sun.nio.fs.WindowsException.translateToIOException(WindowsException.java:92)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:103)
	at java.base/sun.nio.fs.WindowsException.rethrowAsIOException(WindowsException.java:108)
	at java.base/sun.nio.fs.WindowsFileSystemProvider.implDelete(WindowsFileSystemProvider.java:275)
	at java.base/sun.nio.fs.AbstractFileSystemProvider.deleteIfExists(AbstractFileSystemProvider.java:110)
	at java.base/java.nio.file.Files.deleteIfExists(Files.java:1191)
	at java.base/java.nio.file.Files.copy(Files.java:3149)
	at jdk.test.lib.helpers.ClassFileInstaller.writeToDisk(ClassFileInstaller.java:296)
	at jdk.test.lib.helpers.ClassFileInstaller.writeClassToDisk(ClassFileInstaller.java:255)
	at jdk.test.lib.helpers.ClassFileInstaller.writeClassToDisk(ClassFileInstaller.java:237)
	at jdk.test.lib.helpers.ClassFileInstaller.writeClassToDisk(ClassFileInstaller.java:234)
	at jdk.test.lib.helpers.ClassFileInstaller.main(ClassFileInstaller.java:97)
	at compiler.lib.ir_framework.TestFramework.installWhiteBox(TestFramework.java:527)
	... 8 more

```

Weird but reproducible.

**Comment by TobiHartmann** [comment]
Let me try to reproduce with `-DSkipWhiteBoxInstall=true`.

**Comment by TobiHartmann** [comment]
This fixes the issue:
```
- * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI compiler.c2.irTests.TestSuperwordFailsUnrolling
+ * @run main/othervm -Xbootclasspath/a:. -DSkipWhiteBoxInstall=true -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI compiler.c2.irTests.TestSuperwordFailsUnrolling
```

@chhagedorn Maybe the IR verification framework could detect this automatically?

**Comment by chhagedorn** [comment]
> This fixes the issue:
> 
> ```
> - * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI compiler.c2.irTests.TestSuperwordFailsUnrolling
> + * @run main/othervm -Xbootclasspath/a:. -DSkipWhiteBoxInstall=true -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI compiler.c2.irTests.TestSuperwordFailsUnrolling
> ```
> 
> @chhagedorn Maybe the IR verification framework could detect this automatically?

Yes, it would be good to extend the framework to handle this. I added this `-DSkipWhiteBoxInstall` flag at some point to prevent the problem above but was more of a quick fix. I will follow up with an RFE to make this better without the need to use `-DSkipWhiteBoxInstall`.

File: src/hotspot/share/opto/superword.cpp (hunk 2 above)
Comment (rwestrel) [reply src/hotspot/share/opto/superword.cpp]: Should be done in the new commit

**Comment by rwestrel** [comment]
> This fixes the issue:
> 
> ```
> - * @run main/othervm -Xbootclasspath/a:. -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI compiler.c2.irTests.TestSuperwordFailsUnrolling
> + * @run main/othervm -Xbootclasspath/a:. -DSkipWhiteBoxInstall=true -XX:+UnlockDiagnosticVMOptions -XX:+WhiteBoxAPI compiler.c2.irTests.TestSuperwordFailsUnrolling
> ```

New commit has this change

**Comment by rwestrel** [comment]
> That looks good!

Thanks for the review @chhagedorn

**Review by TobiHartmann — APPROVED** [review]
Looks good.

**Pull request closed without merging**